### ISSUE TITLE
Error generating IPA (windows with java 17)

### ISSUE THREAD

**Issue by caraiz** [airsdk/Adobe-Runtime-Support#1764]
Using java 17 version adt generates this error: 

"Cannot sign an IPA file on non-Mac platform: invalid JRE that does not expose sun.security classes"

Tested with Air SDK version: 33.1.1.795

I suppose the problem is the encapsulation of internal packages, and it will fails in next versions too:
https://www.infoq.com/news/2021/06/internals-encapsulated-jdk17

**Comment by ajwfrost** [comment]
Yes exactly .. it looks like it may still be possible to tweak the command lines and change the implementation to use reflection but I suspect that would be a temporary solution. Ultimately, we want to get rid of the use of the internal Java APIs. On MacOS of course we can use the native 'codesign' tool, but there's nothing that really does that on Windows.

We have a task to look further into what could be done here, whether some of the functionality can be reworked using different APIs/tools etc. But for now, please use an earlier Java version, e.g. Java 11 will be supported for another 18 months..

thanks

**Comment by caraiz** [comment]
If this helps, adding "**--add-opens java.base/sun.security.x509=ALL-UNNAMED --add-opens java.base/sun.security.util=ALL-UNNAMED**" in java command  seems solve that problem, but the implementation has changed and generates other error, so really is necessary review the sign process...

'''
Exception in thread "main" java.lang.NoSuchMethodError: 'sun.security.util.ObjectIdentifier sun.security.util.ObjectIdentifier.newInternal(int[])'
        at com.adobe.air.ipa.MachoSigner.signToStream(MachoSigner.java)
        at com.adobe.air.ipa.MachoSigner.newCMSBlob(MachoSigner.java)
        at com.adobe.air.ipa.MachoSigner.sign(MachoSigner.java)
        at com.adobe.air.ipa.MachoSigner.sign(MachoSigner.java)
        at com.adobe.air.ipa.IPASigner.signMacho(IPASigner.java)
        at com.adobe.air.ipa.IPAOutputStream.finalizeSig(IPAOutputStream.java)
        at com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
        at com.adobe.air.ipa.IPAPackager.createPackage(IPAPackager.java)
        at com.adobe.air.ADT.parseArgsAndGo(ADT.java)
        at com.adobe.air.ADT.run(ADT.java)
        at com.adobe.air.ADT.main(ADT.java)
'''

The newInternal method only calls the constructor, so maybe calling directly to the constructor and adding the "--add-opens" can be a initial fix to solve the problem, but how you say, the better is try to avoid all those classes.

Thanks.

**Comment by emmaassin** [comment]
I'm having this issue too.  Is it not currently possible to build an IPA on Windows?

**Comment by FliplineStudios** [comment]
I can still build an IPA on Windows, though I did have to switch to JDK 11.0.13 to get it to work though.

**Comment by emmaassin** [comment]
Switching to JDK 11.0.13 didn't work for me - same error.

**Comment by ajwfrost** [comment]
Java 17 will cause errors but Java 11 should be possible .. although equally @emmaassin, there is a possibility that it's found the Java 17 installation and is still using that. If you enable the logging (airsdk_folder\lib\adt.cfg, to uncomment `DebugOut=true`) you should see more details in the `c:\users\username\adt.log` file..)

thanks

**Comment by shaymolho** [comment]
hey @ajwfrost when is java 17 expected to be supported?

**Comment by cybomon** [comment]
@ajwfrost I'm getting this exact invalid JRE cannot sign error when trying to package an IPA file. I'm using Java 16. Is there a fix for this? I tried compiling using the SDK on my mac and I got an even more intimidating error of various exceptions on compilation.

**Comment by ajwfrost** [comment]
On Windows -> Java 11 is probably still best, I think there are also problems that Android Studio has with the latest Java versions. For IPA files, the Java code is just doing the packaging of the IPA (rather than anything that is inherently Java-ish), the challenge is in the code-signing where on macOS we can just use the native tools, but on Windows there is a whole load of code logic to create the appropriate macOS code signatures - and this was written using the private/internal Java security classes which are no longer accessible in these later JVMs.

So, refusal for the app store wouldn't come from the version of Java we're using, but rather from the execution of our Java app (if that makes sense!) -> you should be okay. But yes, I agree, it would be better if we could avoid the use of these internal classes, we do have an open task for this and we'd looked at it a while back... there may be some other third party tools we can try to pick up for this. I'll see if someone can look into that again with a bit more focus.

thanks

**Comment by cybomon** [comment]
@ajwfrost Thanks for the detailed information. I always did my ipa packaging on windows except for I believe when making an ipa that would run in the xcode simulator. I think I had to use the mac in the past for that. If the mac environment is better for ipa packaging, which would make sense, I'll try to move in that direction. When I have time, I'll look into the error I was getting packaging in osx and try to troubleshoot it. Thanks again.

**Comment by shaymolho** [comment]
@ajwfrost Still cannot build ipa on with java 17 on windows
I don't want to publish this ipa to app store so maybe that could be easier?
the error i'm getting is:

>      [exec] Cannot sign an IPA file on non-Mac platform: invalid JRE that does not expose sun.security classes
[adt.log](https://github.com/airsdk/Adobe-Runtime-Support/files/12387845/adt.log)

**Comment by ajwfrost** [comment]
Yes, I think Java 16 is the latest you can use for signing IPA files on Windows.. Java 11 is still what I tend to use though.

**Comment by shaymolho** [comment]
with java 16 same error:
     [exec] Cannot sign an IPA file on non-Mac platform: invalid JRE that does not expose sun.security classes

**Comment by stephanberger4** [comment]
> Java 17 will cause errors but Java 11 should be possible .. although equally @emmaassin, there is a possibility that it's found the Java 17 installation and is still using that. If you enable the logging (airsdk_folder\lib\adt.cfg, to uncomment `DebugOut=true`) you should see more details in the `c:\users\username\adt.log` file..)
> 
> thanks

Hi. I've got the same problem. I did the uncomment and saw the details in the adt.log file. I saw that it was, indeed, still using the 17 version of java instead of the 16 that I have installed. ("Cannot create sun security class: JRE version = 17.0.6+9-LTS-190, Java Home = C:\Program Files\Adobe\Adobe Animate 2024\jre"). 

How/where can I tell it to use java 16 instead of 17 ?

Thx
